### PR TITLE
feat: kona-providers

### PR DESCRIPTION
**Pull request by refcell**
### Description

Splits out provider traits from `kona-derive` into a new minimal crate that can be implemented by various provider implementation crates like `kona-providers-alloy`.

Also bumps `alloy` and `op-alloy` dependencies to use upstream block types.

### CODE REVIEW

**Comment by refcell** [comment]
* **#596** <a href="https://app.graphite.dev/github/pr/anton-rs/kona/596?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @refcell and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: crates/providers/src/utils.rs
Comment (refcell) [inline crates/providers/src/utils.rs:1]: Panning to upstream these utilities to `op-alloy-consensus`, next to the block type. But hoisting here for now to sprint to green CI

File: crates/derive/src/attributes/mod.rs [hunk 1]
```diff
@@ -97,7 +98,11 @@ where
                     .await
                     .map_err(|e| PipelineError::BadEncoding(e).crit())?;
             sys_config
-                .update_with_receipts(&receipts, &self.rollup_cfg, header.timestamp)
+                .update_with_receipts(
+                    &receipts,
+                    self.rollup_cfg.l1_system_config_address,
+                    self.rollup_cfg.is_ecotone_active(header.timestamp),
```
Comment (clabby) [inline crates/derive/src/attributes/mod.rs:104]: I'm not a huge fan of this API - we should just pass the rollup cfg ya? Op-alloy has it

File: crates/derive/src/batch/span_batch/batch.rs [hunk 2]
```diff
@@ -288,25 +286,33 @@ impl SpanBatch {
...
                     );
                     return BatchValidity::Drop;
                 }
-                for j in 0..batch_txs.len() {
+                let batch_txs_len = batch_txs.len();
+                #[allow(clippy::needless_range_loop)]
+                for j in 0..batch_txs_len {
                     let mut buf = Vec::new();
-                    safe_block_txs[j + deposit_count].encode_2718(&mut buf);
+                    safe_block_txs.transactions[j + deposit_count].encode_2718(&mut buf);
```
Comment (clabby) [inline crates/derive/src/batch/span_batch/batch.rs:306]: Nit on var name, should just be `safe_block`

**Review by clabby — APPROVED** [review]

**Pull request merged**